### PR TITLE
Update provisioning and build scripts to VS2017 (as per #4151)

### PR DESCRIPTION
**Pull request by momopranto**
[#4151](https://github.com/facebook/osquery/issues/4151)

Dev env provisioning script now installs Visual Studio 2017 now by default. Original script that provisions VS2015 env is still available in same directory. Build script also now tries to build with VS2017 but will use VS2015 as backup. 

vsdeploy.xml is no longer used by default provisioning script but is kept for the VS2015 script. vsinstall.json specifies custom VS installation components.

Also suppressed Boost config warnings.

### CODE REVIEW

**Comment by osqueryer** [comment]
:-1: The commit 93254b2fe367f48135f47cb354f2faf3e5bafb09 (Job results: [7374](https://jenkins.osquery.io/job/osqueryPullRequestLinux/7374/)) failed one or more tests (**Linux**).

**Comment by osqueryer** [comment]
:-1: The commit 93254b2fe367f48135f47cb354f2faf3e5bafb09 (Job results: [4279](https://jenkins.osquery.io/job/osqueryPullRequestWindows/4279/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 93254b2fe367f48135f47cb354f2faf3e5bafb09 (Job results: [4282](https://jenkins.osquery.io/job/osqueryPullRequestWindows/4282/)) failed one or more tests (**Windows**).

**Comment by muffins** [comment]
ok to test

**Comment by osqueryer** [comment]
:-1: The commit 93254b2fe367f48135f47cb354f2faf3e5bafb09 (Job results: [4283](https://jenkins.osquery.io/job/osqueryPullRequestWindows/4283/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 93254b2fe367f48135f47cb354f2faf3e5bafb09 (Job results: [4284](https://jenkins.osquery.io/job/osqueryPullRequestWindows/4284/)) failed one or more tests (**Windows**).

**Comment by muffins** [comment]
Windows somehow grabbed the release tests I was drafting and seems to be broken. AMIs have been recreated, going to re-test to see if we're working again.

**Comment by muffins** [comment]
ok to test

**Comment by osqueryer** [comment]
:-1: The commit 93254b2fe367f48135f47cb354f2faf3e5bafb09 (Job results: [4285](https://jenkins.osquery.io/job/osqueryPullRequestWindows/4285/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 93254b2fe367f48135f47cb354f2faf3e5bafb09 (Job results: [4286](https://jenkins.osquery.io/job/osqueryPullRequestWindows/4286/)) failed one or more tests (**Windows**).

File: CMakeLists.txt [hunk 1]
```diff
@@ -603,6 +604,9 @@ else()
 endif()
 
 if(WINDOWS)
+  if(NOT OS_LIB_DIRS)
+    set(OS_LIB_DIRS)
```
Comment (muffins) [inline CMakeLists.txt]: What's going on here? Why is this being set?

File: tools/build.ps1 [hunk 2]
```diff
@@ -17,27 +17,9 @@ if (-not (Test-Path $utils)) {
...
-      $vsVersion = $toks[1].Split(".")[0]
-    }
-    if ($toks[0].trim() -like 'installationPath') {
-      $vsLoc = [System.String]::Join(":", $toks[1..$toks.Length])
-    }
-  }
-  $vsLoc = $vsLoc.trim()
-  $vsVersion = $vsVersion.trim()
+  $vsinfo = Get-VSInfo
+  $vsLoc = $($vsinfo.location)
```
Comment (muffins) [inline tools/build.ps1]: You shouldn't need to add the extra `$()` wrap for this, just access the variable directly `$vsLoc = $vsinfo.location`

File: tools/build.ps1 [hunk 3]
```diff
@@ -17,27 +17,9 @@ if (-not (Test-Path $utils)) {
...
+  $vsVersion = $($vsinfo.version)
```
Comment (muffins) [inline tools/build.ps1]: same.

File: tools/build.ps1 [hunk 4]
```diff
@@ -67,12 +49,20 @@ function Invoke-VcVarsAll {
...
 # A helper function to call CMake and generate our solution file
 function Invoke-OsqueryCmake {
+  $vsinfo = Get-VSInfo
   $cmake = (Get-Command 'cmake').Source
-  $cmakeArgs = @(
-    '-G "Visual Studio 14 2015 Win64"',
-    '-T v140',
-    '../../'
-  )
+  if ($($vsinfo.version) -eq '15'){
```
Comment (muffins) [inline tools/build.ps1]: Same here, just access directly.

File: tools/provision-vs2015.ps1 [hunk 5]
```diff
@@ -0,0 +1,507 @@
+#  Copyright (c) 2014-present, Facebook, Inc.
```
Comment (muffins) [inline tools/provision-vs2015.ps1]: This is a _massive_ amount of code duplication. What components are you needing out of this? We should take whatever functionality is needed here, and fold this into the generic `provision.ps1`, _or_ dot source whatever functionality is needed from `provision.ps1` and keep all of our provisioning logic there. In this case I'd say let's stick with the former.

File: tools/provision/chocolatey/osquery_utils.ps1 [hunk 6]
```diff
@@ -225,3 +225,39 @@ function Start-OsqueryProcess {
...
 }
+
+# A helper function to derive the latest VS install
+function Get-VSInfo {
+
+  # Attempt to make use of vswhere to derive the build tools scripts
+  $vswhere = (Get-Command 'vswhere').Source
+  $vswhereArgs = @('-latest', '-legacy')
+  $vswhereOut = (Start-OsqueryProcess $vswhere $vswhereArgs).stdout
+  $vsinfo = "" | Select-Object -Property location,version
```
Comment (muffins) [inline tools/provision/chocolatey/osquery_utils.ps1]: This is a pretty sweet hack! However let's be more explicit in our object creation:
```
$vsinfo = New-Object -TypeName psobject
$vsinfo | Add-Member -MemberType NoteProperty -Name version -Value ''
$vsinfo | Add-Member -MemberType NoteProperty -Name location -Value ''
```

**Comment by muffins** [comment]
ok to test

File: tools/provision/chocolatey/osquery_utils.ps1 [hunk 7]
```diff
@@ -225,3 +225,39 @@ function Start-OsqueryProcess {
...
+      $vsinfo.location = [System.String]::Join(":", $toks[1..$toks.Length])
+    }
+  }
+  $vsinfo.location = $vsinfo.location.trim()
+  $vsinfo.version = $vsinfo.version.trim()
+  return $vsinfo
+}
+
+# A helper function to determine if VS 2017 is installed
+function Check-VSInstall {
```
Comment (muffins) [inline tools/provision/chocolatey/osquery_utils.ps1]: Let's maybe change this verbage to be specifically what we're getting back, something like `Get-VisualStudio2017Installed`?

**Comment by osqueryer** [comment]
:-1: The commit 93254b2fe367f48135f47cb354f2faf3e5bafb09 (Job results: [4295](https://jenkins.osquery.io/job/osqueryPullRequestWindows/4295/)) failed one or more tests (**Windows**).

**Comment by osqueryer** [comment]
:-1: The commit 93254b2fe367f48135f47cb354f2faf3e5bafb09 (Job results: [4296](https://jenkins.osquery.io/job/osqueryPullRequestWindows/4296/)) failed one or more tests (**Windows**).

**Comment by muffins** [comment]
ok to test

File: CMakeLists.txt (hunk 1 above)
Comment (momopranto) [reply CMakeLists.txt]: Handling OS_LIB_DIRS not defined for the removed duplicates macro

Comment (muffins) [reply CMakeLists.txt]: I'm not 100% convinced that this is needed. If you remove it do builds break? I tested this on a VM and it seemed like it was building just fine. If we need it that's fine, but would you mind finding some links or more context around why this is needed? As it stands we're just adding yet another build definition, but it's not clear _who_ needs this.

**Pull request merged**